asdf_zarr/storage.py

```diff
@@ -156,6 +157,42 @@ def __init__(self, ctx, chunk_block_map_index, zarray_meta):
             self._chunk_asdf_keys[chunk_key] = asdf_key
             self._chunk_callbacks[chunk_key] = ctx.get_block_data_callback(block_index, asdf_key)
 
+    def __getstate__(self):
+        state = {}
+        state["_sep"] = self._sep
+        if hasattr(self, "_chunk_info"):
+            # handle instance that was already pickled and unpickled
+            state["_chunk_info"] = self._chunk_info
+        else:
+            # and instance that was not yet pickled
+
+            # for each callback, get the file uri and block offset
+            def _callback_info(cb):
+                return {
+                    "offset": cb(_attr="offset"),
+                    "uri": cb(_attr="_fd")().uri,
```

Comment:
The private `_attr` argument is used here (as is needed for ndarray as [noted in the code](https://github.com/asdf-format/asdf/blob/74893af97bee97a2d66dcd9e0147736761a8b6ff/asdf/_block/callback.py#L11-L15)). Perhaps this can inform what a "low level" block API would look like. By using the `offset` and the `uri` this will not work (and likely produce junk data) after an `update` of the backing ASDF file. Minimally we should document this danger and it may be one better to check the file mode and not allow pickling when the file is opened 'rw'.